keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1806 @@
+this KEP, the PodSpec is extended to include pod-level resource requests and
+limits. These values will be directly used to configure the pod-level cgroup,
+providing a more direct and explicit way to manage the overall resource
+consumption of the pod.
+
+#### PodSpec API changes
+
+New field in `PodSpec`
+
+```
+type PodSpec struct {
+...
+  // The total amount of compute resources required by all containers in the pod.
+  // Currently, only CPU and Memory resources are supported .
+  // +featureGate=PodLevelResources
+  // +optional
+  Resources ResourceRequirements
+}
+```
+
+#### PodSpec Validation Rules
+
+##### Proposed Validation Rules
+
+To ensure clarity and prevent ambiguity in resource allocation, the following
+validation rules are proposed for pod-level resource specifications:
+* Mandatory Requests: The introduction of default values for container requests in Kubernetes
+  [PR#12035](https://github.com/kubernetes/kubernetes/pull/12035) was primarily
+  driven by the need for backward compatibility when requests were added to the
+  Kubernetes. Since limits already existed, defaulting those to limits or to a
+  minimum value ensured that existing pods continue to function as expected.
+  However, this approach led to guessing user intention and use of some
+  [arbitrary minimum
+  values](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/util/pod_resources.go#L32-L37).
+  This KEP for pod-level resources aims to address those limitations by
+  encouraging users to specify pod-level resource request in 
+  following ways to avoid reliance on potentially unsuitable default
+  values:
+
+    * Explicitly Defined: Users can explicitly set pod-level requests in the pod spec.  
+    * Derived From Containers: If no pod-level limit is defined, Kubernetes can
+      derive it from container-level requests. Users should ensure that at least
+      one container, including init containers, regular containers, or sidecars
+      (including those injected later), has a request specified for this work
+      effectively.
+    * Combined Approach: Users can specify both container-level and pod-level
+      requests. This can be useful for fine-tuning the priority order in OOM
+      killing scenarios, as the oom_score_adj calculation will take both
+      pod-level and container-level requests into account. (More details about
+      oom_score_adj in #oom-score-adjustment). The scheduler will however use
+      pod-level request, in this case, to find a suitable node.
+
+  This ensures that Kubernetes can determine a minimum resource requirement for
+  the pod, either directly or indirectly from the container requests (using the
+  formula in [KEP
+  #753](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/753-sidecar-containers/READMEmd#resources-calculation-for-scheduling-and-pod-admission))
+  
+  **Rule: If no requests are provided either at the pod level or at the container level,
```

Comment:
This sounds like a breaking change to the existing best effort behavior assuming no pod level requests are set.

Reply:
Correct, I will discuss and change this. Today we assume arbitrary container level minimum values. Do you have suggestions for what the minimum at pod level could be? No. of containers * the minimum values that we have today? 

These arbitrary values that we have today, seem very high for mininum resource requests… and they are used only by the scheduler to find a suitable node.

https://github.com/kubernetes/kubernetes/blob/60efdae3d1c4f5ac7e3aa9acbfa8a818e1fb2093/pkg/scheduler/util/pod_resources.go#L36